src/udp_tracker.rs

```diff
@@ -0,0 +1,311 @@
+use crate::common::*;
+
+#[derive(Clone, Copy, Debug)]
+struct ConnectRequest {
+  protocol_id: u64,
+  action: u32,
+  transaction_id: u32,
+}
+
+#[derive(Debug)]
+struct ConnectResponse {
+  action: u32,
+  transaction_id: u32,
+  connection_id: u64,
+}
+
+#[derive(Debug)]
+struct AnnounceRequest {
+  connection_id: u64,
+  action: u32,
+  transaction_id: u32,
+  infohash: [u8; 20], // Infohash,
+  peer_id: [u8; 20],
+  downloaded: u64,
+  left: u64,
+  uploaded: u64,
+  event: u64,
+  ip_address: u32,
+  num_want: u32,
+  port: u16,
+}
+
+#[derive(Debug)]
+struct AnnounceResponse {
+  action: u32,
+  transaction_id: u32,
+  interval: u32,
+  leechers: u32,
+  seeders: u32,
+}
+
+trait Request {
+  type Response: Response;
+
+  // It seems like there're many ways to marshal these messages. I'd prefer
+  // something readable and fast. Here are some options:
+  //
+  // 1. Cursor. With multiple write calls. ◆
+  //
+  // 2. IoSlice. Is there overhead to all the ::new() calls? This will probably be
+  //    the winner once the write_vectored() calls diffuse out to std::net
+  //
+  // 3. Serde. (Already a dependency). Seems like overkill.
+  //
+  // 4. Binread.
```

Comment:
I don't have any particularly good suggestions for which crate to use. Binread looks really cool, but doesn't support serialization. I think multiple write calls might be okay for now, if there isn't a clear winner otherwise.

Reply:
I still reckon vectored writes will eventually be the best way to do this.

Reply:
I think at the moment we shouldn't worry about performance at all. With single announces, any performance improvements will be unnoticeable to end users. For simplicity, I'd just do extending vectors with byte slices, and then at the end write the whole vector. Then later if performance is an issue, we can profile and improve it.